docs/proposals-accepted/202205-vertical-query-sharding.md

```diff
@@ -0,0 +1,152 @@
+
+### Initiating sharded queries
+
+The Query Frontend component already has useful splitting and merging capabilities which are currently used for sharding queries horizontally. In order to provide the best possible user experience and not burden users with running additional component(s), we propose adding a new middleware to Query Frontend which will implement the vertical sharding algorithm proposed here after the step-alignment and horizontal time-slicing steps. This would allow users to restrict the maximum complexity of a given query based on vertical sharding of already time-sliced queries. The only new user-specified parameter would be the number of shards in which to split PromQL aggregations.
+
+Integrating vertical-query sharding in QFE also has the added benefits of: 
+* Utilizing the existing query range caching implementation 
+* Vertical-query sharding a query that has already been time-sliced reduces the cardinality of each respective query shard
+The following diagram illustrates how the sharding algorithm would work end to end:
+
+![Vertical sharding](../img/vertical-sharding.png "Vertical query sharding")
+
+### Drawbacks & Exceptions
+
+*Not all queries are easily shardeable*
+There are certain aggregations for which sharding cannot be performed safely. For these cases the sharder can simply fall back to completely executing the expression as a single query. These cases include the use of functions such as `label_replace` and `label_join` which create new labels inside PromQL, when the query is being executed. Since such labels can be arbitrarily created while executing queries, Stores will be unaware of them and cannot take them into account when sharding the matching series set. 
+
+#TODO Mention promql parsing/transformation to increase queries that can be sharded
+
+*Impact of sharded queries on block retrieval* 
+Given that queries are sharded based on series, and not time, for a given leaf most metrics for each independent shard will likely reside in the same block(s). This can multiply the cost of retrieving series from a given block since the block has to be traversed N number of times, where N is the number of distinct shards.
+
+In our sharded query benchmarks, we found that the increased volume of Series calls to Store Gateways and Receivers did not lead to a significant latency increase. However, we did notice a higher remote-read latency in Prometheus caused by multiple remote-read calls for the same Series call. This is due to the fact that the sharding has to be done in the sidecar, after retrieving all matching series for a query. A similar concern was raised in Prometheus already, with a proposal to support sharding series natively in the Prometheus TSDB (https://github.com/prometheus/prometheus/issues/10420). 
```

Comment:
I think we'll have to take care of respecting `max-query-parallelism`, but other than that, we should be able to use both vertical and horizontal sharding independently of each other.

Reply:
The proposed implementation does vertical sharding after time slicing: 
`step-align -> timeslice -> vertically shard query` 

Both techniques are complimentary as the maximum query cardinality becomes `active series per configured QFE time slice (e.g. 12 hours) / number of shards configured (e.g. 10 vertical shards)`

Reply:
As mentioned in https://cloud-native.slack.com/archives/CL25937SP/p1648741512490379, I think we can reuse the Mimir Prometheus fork.
I am wondering if we can try to integrate with that first. It should be more performant in the case of the sidecar and an additional label matcher is easy to add at the query time.

Reply:
I think we should clarify in the proposal that a high cardinality query would still execute faster overall, just that Prometheus will take a bit longer to retrieve requested series over remote-read. 

Could you elaborate a bit on how we would reuse the Mimir Prometheus fork? Will people have to run that instead of upstream Prometheus?

Reply:
That's a good point. We cannot force users to run that version of Prometheus. But I believe that the upstream TSDB will have that native sharding support later. I hope our solution can be at least compatible with it and later we can switch to it once it is upstreamed.

Reply:
I just saw that we can also request only series without chunks from remote read: https://github.com/prometheus/prometheus/blob/3947238ce0e3ff7e7c6e9502f73275cdbc233054/tsdb/querier.go#L140

Something I would also like to try out in the sidecar is:
* request only series from remote read
* calculate only the series which match a shard
* request series and chunks for a shard

I'll report findings once I've run some benchmarks.

Reply:
This one works but it sends more remote read requests to Prometheus (one request per series). I am curious what's the benchmark results...

Reply:
I was thinking we could fetch batches of series in fewer requests, but now that I'm looking at the code, I don't think that is easy to do. So this approach probably won't work as I expected.